### ISSUE TITLE
Being able to run multiple providers simultaneously 

### ISSUE THREAD

**Issue by ainsey11** [pier-oliviert/phonebook#11]
At the moment, we listen to all CRD's generated with the Kind of DNSRecord which is 100% fine,

However, we don't currently implement any filtering, to ensure that the domain specified in the record, is one that provider can handle

For example, my home network

I have a bind9 DNS cluster I wish to update over RFC2136
I also have an external facing domain, I wish to update with the Azure DNS provider
Finally, I have some cloudflare tunnels, and a CF DNS account I'd like to update with the CF DNS Provider
and so on...

It would be good to be able to specify which zone/domain is for which provider, and be able to run multiple configurations, or multiple instances of the same provider but different accounts / dns hosts. At the moment this could be quite tricky to implement, and I'm not sure where I'd start 

Just wanted to open it up for discussion on how we could tackle this - it does also creep a bit into one huge downside to external-dns too, which is the ability to handle split-horizon DNS, which technically this would allow to work (2 CRDs for the same domain, backed by different providers and a different IP)

**Comment by pier-oliviert** [comment]
Glad you brought this up, this was one feature I wanted to implement but waited until someone needed it. To get the conversation started, here's what I had in mind initially:

In the current form the `DNSRecord` CRD has the following structure
```
type DNSRecordSpec struct {
	Zone string `json:"zone"`
	RecordType string `json:"recordType"`
	Name string `json:"name"`
	Targets []string `json:"targets"`
	Properties map[string]string `json:"properties,omitempty"`
	TTL *int64 `json:"ttl,omitempty"
}
```

What I wanted to do was to add an **optional** field called `Provider *string` which would reference a new CRD: `DNSProvider`. This would mean that providers wouldn't be created at runtime but rather be a cluster-scoped CRD that can be defined ad-hoc.

The [`DefaultProvider()`](https://github.com/pier-oliviert/phonebook/blob/main/cmd/main.go#L158) was always intended to be a fallback provider if the optional `Provider` field wasn't set on `DNSRecord`.

The CRD could look something like this:

```
type DNSProviderSpec struct {
        // Name is the value you use in DNSRecordSpec for the `Provider *string` optional field
	Name string `json:"name"`
	
	// Zone it has authority over, the DNSRecord needs to match the Provider & the Zone
	Zone string `json:"zone"`
	
	// Configuration Secret Ref for the Provider's 
	SecretRef types.LocalNamespace `json:"secretRef,omitempty"`
	
	// Non Secret configuration options
	Config map[string]string `json:"config,omitempty"`
}
``` 

Now the Reconciliation process for the DNSProvider would be quite simple, anytime a change happens to one of those resources, the controller would create/update/remove the Provider and store it in-memory.

When a DNSRecord gets created/deleted, Phonebook wouldn't have to fetch those CRD but would rather just pull the provider from the in-memory list it has and find the Provider that matches the spec. If none is found, an error is returned, and the error is then set on the DNSRecord's EventRecorder as well as in the ConditionProvider.

What do you think?

**Comment by ainsey11** [comment]
I like the approach, means we can keep the existing functionality for an easy out-of-the-box expirience setting things up in values.yaml, and if users want more providers they can just add them on via the CRD,

My only other fleeting thought is - do we not just do this for everything?
we could have the helm chart deploy the `DNSProvider` CRD based on the items in the values.yaml, we could even use helms `loop` ability to allow the user to configure multiple providers from the `values.yaml` file,

My logic with the latter is that it means users can change providers, without having to remove the operator and re-install / re-configure / re-run the helm deploy. It's quite a big change in behaviour, but when I've built other operators this seems to be the pattern most go for (and I followed)

It's a tricky one, be interested to see how you feel about that idea

**Comment by pier-oliviert** [comment]
> My only other fleeting thought is - do we not just do this for everything?
we could have the helm chart deploy the DNSProvider CRD based on the items in the values.yaml

I like this. I'll try to build a PoC in a PR so we can discuss it further.

**Comment by WladyX** [comment]
Also looking forward for this one.
My domain has 2 Authoritative DNS (desec + gcore) so I need to update both with the same records.
Currently I'm using dnscontrol but would like to switch to k8s.
Thanks!

**Comment by pier-oliviert** [comment]
Just quickly noting here that I'm working on this, I will have more to share very soon. I'm very excited about it, but I need a little bit more time before I can share it with ya'll.

**Comment by pier-oliviert** [comment]
With #13 merged and the Release of 0.3.4 I believe I can close this as multiple providers is now supported.

I'm still working on more documentation on how to create a new provider, etc. But that shouldn't prevent anyone from using the ones that already exists.
